### ISSUE TITLE
Feature Request: Create Optional Field for Related Model

### ISSUE THREAD

**Issue by cqh963852** [chrishoermann/zod-prisma-types#61]
![image](https://user-images.githubusercontent.com/17702287/214572967-1199081d-dea9-4472-8f87-1931de5052e3.png)

I am using model schema serialize data in fastify.

Prisma's data is serialized faster after using the serialization function built by schema.

Is there any way I can add the optional schema?

**Comment by chrishoermann** [comment]
@cqh963852 in the recently released version `2.0.0-beta.0`  I changed the implementation to generate a separate schema that has  all the relations. The inline method did not work that well because of the recursiveness of the relation fields. For recursive fields a `z.ZodType<myType>` annotation is necessary to enable zod to reference itself. But this means that the schema looses he possibility of using methods like `merge`, `omit`, `pick`, etc. that the simple object schema provides.

```prisma
generator zod {
  // ...rest of config
  createRelationValuesTypes = true
}

  id         String      @id @default(cuid()) 
  email      String      @unique 
  name       String?
  posts      Post[]
  profile    Profile?
  role       Role[]      @default([USER, ADMIN])
  scalarList String[]

  lat Float
  lng Float

  location Location? @relation(fields: [lat, lng], references: [lat, lng])
}

model Post {
  id             Int           @id @default(autoincrement())
  title          String
  content        String?
  published      Boolean       @default(false)
  author         User          @relation(fields: [authorId], references: [id])
  authorId       String
  jsonOpt        Json?
  anotherJsonOpt Json?
  anotherEnum    AnotherEnum[]
}

model Profile {
  id     Int        @id @default(autoincrement())
  bio    String
  user   User       @relation(fields: [userId], references: [id])
  userId String     @unique
  role   Role[]     @default([USER, ADMIN])
  second SecondEnum @default(ONE)
}

model Location {
  lat  Float
  lng  Float
  User User[]

  @@id([lat, lng])
}
```

The above model would generate the following model schemas:

```ts
// USER
//------------------------------------------------------


export const UserSchema = z.object({
  role: RoleSchema.array(),
  enum: AnotherEnumSchema,
  id: z.string().cuid(),
  email: z.string().email({ message: "Invalid email address" }),
  /**
   * some other comment
   * some message after
   */
  name: z.string().min(1).max(100).nullish(),
  scalarList: z.string().array(),
  lat: z.number(),
  lng: z.number(),
})

export const UserOptionalDefaultsSchema =UserSchema.merge(z.object({
  role: RoleSchema.array().optional(),
  enum: AnotherEnumSchema.optional(),
  id: z.string().cuid().optional(),
}))

export type UserWithRelations = z.infer<typeof UserSchema> & {
  posts: PostWithRelations[];
  profile?: ProfileWithRelations | null;
  location?: LocationWithRelations | null;
}

export const UserWithRelationsSchema: z.ZodType<UserWithRelations> = UserSchema.merge(z.object({
  posts: z.lazy(() => PostWithRelationsSchema).array(),
  profile: z.lazy(() => ProfileWithRelationsSchema).nullish(),
  location: z.lazy(() => LocationWithRelationsSchema).nullish(),
}))

// POST
//------------------------------------------------------


export const PostSchema = z.object({
  anotherEnum: AnotherEnumSchema.array(),
  id: z.number().int(),
  title: z.string(),
  content: z.string().nullish(),
  published: z.boolean(),
  authorId: z.string(),
})

export const PostOptionalDefaultsSchema =PostSchema.merge(z.object({
  id: z.number().int().optional(),
  published: z.boolean().optional(),
}))

export type PostWithRelations = z.infer<typeof PostSchema> & {
  author: UserWithRelations;
}

export const PostWithRelationsSchema: z.ZodType<PostWithRelations> = PostSchema.merge(z.object({
  author: z.lazy(() => UserWithRelationsSchema),
}))

// PROFILE
//------------------------------------------------------


export const ProfileSchema = z.object({
  role: RoleSchema.array(),
  second: SecondEnumSchema,
  id: z.number().int(),
  bio: z.string(),
  userId: z.string(),
})

export const ProfileOptionalDefaultsSchema =ProfileSchema.merge(z.object({
  role: RoleSchema.array().optional(),
  second: SecondEnumSchema.optional(),
  id: z.number().int().optional(),
}))

export type ProfileWithRelations = z.infer<typeof ProfileSchema> & {
  user: UserWithRelations;
}

export const ProfileWithRelationsSchema: z.ZodType<ProfileWithRelations> = ProfileSchema.merge(z.object({
  user: z.lazy(() => UserWithRelationsSchema),
}))

// LOCATION
//------------------------------------------------------


export const LocationSchema = z.object({
  lat: z.number(),
  lng: z.number(),
})

export type LocationWithRelations = z.infer<typeof LocationSchema> & {
  User: UserWithRelations[];
}

export const LocationWithRelationsSchema: z.ZodType<LocationWithRelations> = LocationSchema.merge(z.object({
  User: z.lazy(() => UserWithRelationsSchema).array(),
}))
```

**Comment by chrishoermann** [comment]
closing since this feature is implemented in the lates official version
